tools/build-tools/src/typeValidator/interfaceValidator.ts

```diff
@@ -0,0 +1,307 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { DiagnosticCategory, InterfaceDeclaration, Node, Project, TypeChecker } from "ts-morph";
+import {
+    getCallSignatureReplacement,
+    getIndexSignatureReplacement,
+    getMethodReplacement,
+    getPropertyReplacement,
+} from "./memberDecomposition";
+import { decomposeType, GenericsInfo, typeToString } from "./typeDecomposition";
+import { BreakingIncrement, IValidator, log } from "./validatorUtils";
+
+/**
+ * Total result of a class decomposition which may be reconstructed into an equivalent class
+ * declaration to bypass issues with normal class type comparisions
+ */
+interface InterfaceData {
+    readonly name: string;
+    readonly typeParameters: string[];
+    readonly properties: string[];
+    readonly replacedTypes: Set<string>;
+    readonly requiredGenerics: GenericsInfo;
+}
+
+function mergeIntoSet<T>(into: Set<T>, from: Set<T>) {
+    from.forEach((v) => into.add(v));
+}
+
+/**
+ * A decent chunk of this is similar/identical to ClassValidator and could maybe be refactored
+ */
+export class InterfaceValidator implements IValidator {
+    private oldTypeData?: InterfaceData;
+    private newTypeData?: InterfaceData;
+
+    public decomposeDeclarations(
+        oldTypeChecker: TypeChecker,
+        oldDecl: InterfaceDeclaration,
+        newTypeChecker: TypeChecker,
+        newDecl: InterfaceDeclaration,
+    ) {
+        this.oldTypeData = this.decompose(oldTypeChecker, oldDecl);
+        this.newTypeData = this.decompose(newTypeChecker, newDecl);
+    }
+
+    /**
+     * Break down a normal interface declaration into all its parts to facilitate type comparision
+     * - Remove external type dependencies to examine separately
+     * - Do not consider declaration merging
+     *
+     * - Do not consider unusual interactions with inheritance.  E.g. InterfaceA extends ClassA, and
+     *   ClassB implements InterfaceA but does not extend ClassA.  If a private property myPrivateProp
+     *   is added on ClassA, ClassB will no longer be valid because it will be either missing a
+     *   declaration for myPrivateProp or its declaration of myPrivateProp will be considered separate
+     *   from that of ClassA's because it does not directly extend ClassA.
+     *
+     * @param typeChecker - The TypeChecker object from the node's TS project for getting type names
+     * @param node - The interface declaration node to decompose
+     * @returns - InterfaceData for the decomposed interface declaration
+     */
```

Comment:
```suggestion
     * Break down a normal interface declaration into all its parts to facilitate type comparison
     *
     * - Remove external type dependencies to examine separately
     * - Do not consider declaration merging
     *
     * - Do not consider unusual interactions with inheritance.  E.g. InterfaceA extends ClassA, and
     *   ClassB implements InterfaceA but does not extend ClassA.  If a private property myPrivateProp
     *   is added on ClassA, ClassB will no longer be valid because it will be either missing a
     *   declaration for myPrivateProp or its declaration of myPrivateProp will be considered separate
     *   from that of ClassA's because it does not directly extend ClassA.
     *
     * @param typeChecker - The TypeChecker object from the node's TS project for getting type names
     * @param node - The interface declaration node to decompose
     * @returns - InterfaceData for the decomposed interface declaration
     */
```